### PR TITLE
Add wreply support in sign out endpoint

### PR DESCRIPTION
**Pull request by vadimi**
Add PostLogoutRedirectUris per relying party in order to validate them properly.

I basically took @gnjack idea from PR #44, but made PostLogoutRedirectUris relying party specific, so we can validate them properly by using `wtrealm` parameter. I also added some more validation and logging. It should help with issue #29.

### CODE REVIEW

**Comment by vadimi** [comment]
@leastprivilege any chance you can take a look at that PR? Thank you.

**Comment by leastprivilege** [comment]
Hey - sorry very busy right now. It's on my TODO.

**Comment by gnjack** [comment]
Nice work, thanks for picking up my abandoned PR!

One thing I noticed is that you have made the `wtrealm` parameter required on all sign out requests.
This will probably break compatibility with most relying party applications because `wtrealm` is optional on a `wasignout1.0`, and isn't even defined in [the ws-federation spec for the sign-out message syntax](http://docs.oasis-open.org/wsfed/federation/v1.2/ws-federation.html#_Toc42337314). Although it is [used in this way by Azure ACS](https://msdn.microsoft.com/en-us/library/azure/dn223670.aspx?f=255&MSPPError=-2147217396).

I would suggest that when wtrealm is missing, retain the current behaviour by moving the check to `ProcessSignOutAsync`:
```
if (String.IsNullOrWhiteSpace(msg.Reply) || String.IsNullOrWhiteSpace(msg.GetParameter("wtrealm")))
{
    return RedirectToLogOut();
}
```

This will result in a dead end signed out page, instead of returning a `SignOutValidationResult` with `Error = "invalid_relying_party"` which results in an error page `return BadRequest(result.Error);`

**Comment by vadimi** [comment]
Ah, good catch. Thank you @gnjack

File: source/WsFederationPlugin/WsFederationController.cs [hunk 1]
```diff
@@ -38,29 +39,39 @@ namespace IdentityServer3.WsFederation
...
             _signInResponseGenerator = signInResponseGenerator;
             _metadataResponseGenerator = metadataResponseGenerator;
             _cookies = cookies;
             _wsFedOptions = wsFedOptions;
+            _redirectUriValidator = redirectUriValidator;
+            _signOutValidator = signOutValidator;
         }
 
         [Route("")]
-        [SecurityHeaders(EnableCsp = false, EnableXfo = false)]
```
Comment (scottbrady91) [inline source/WsFederationPlugin/WsFederationController.cs]: Was this attribute removed for a reason?

**Comment by vadimi** [comment]
@scottbrady91 thanks, fixed

**Comment by scottbrady91** [comment]
@brockallen Dom mentioned this might affect the logout/cleanup mechanism in core. Otherwise the PR works fine.

**Comment by brockallen** [comment]
No idea -- I've have to sit down and go thru it. I'll try to get to it as soon as I can.

**Comment by vadimi** [comment]
hey @brockallen. any updates on that?

**Comment by onovotny** [comment]
Would love to have this

**Comment by brockallen** [comment]
@vadimi no, sorry. this was just brought back to my attention.... it's on my TODO for tomorrow AM

**Comment by brockallen** [comment]
This all seems to work well (including the SLO iframe). 

Only issue I have is that on an invalid wreply we're just issuing a 400 and the end-user doesn't get the best experience. Perhaps we can/should trigger an error page instead? We don't currently have an OWIN extension to render the error page, but perhaps we should.

**Comment by brockallen** [comment]
I'll merge for now, and we can discuss what to do for error pages.

**Comment by vadimi** [comment]
great, thanks a lot!

**Comment by onovotny** [comment]
So my only remaining question is where we should pull this from? Are you going to release an updated NuGet soon? Should we pull from MyGet?

**Comment by brockallen** [comment]
It's now released on nuget

**Pull request merged**